### PR TITLE
DAOS-13801 array: fix possible double free with FI

### PR DESCRIPTION
**Pull request by mchaarawi**
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by ashleypittman** [comment]
FYI - I'm not seeing a double free or assertion failure with this patch but I do still see one memory leak of an array pointer but that may be from a different function.

**Review by ashleypittman — DISMISSED** [review]
This fixes the issue as reported however there remain other problems with the truncate code.

**Review by ashleypittman — APPROVED** [review]

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -1246,8 +1246,14 @@ open_file(dfs_t *dfs, dfs_obj_t *parent, int flags, daos_oclass_id_t cid,
 		rc = insert_entry(dfs->layout_v, parent->oh, DAOS_TX_NONE, file->name, len,
 				  DAOS_COND_DKEY_INSERT, entry);
 		if (rc == EEXIST && !oexcl) {
+			int rc2;
+
 			/** just try fetching entry to open the file */
-			daos_array_close(file->oh, NULL);
+			rc2 = daos_array_close(file->oh, NULL);
+			if (rc2 == -DER_NOMEM)
+				rc2 = daos_array_close(file->oh, NULL);
```
Comment (wangshilong) [inline src/client/dfs/dfs.c:1254]: Just curious calling this twice for ENOME is to fix CI memory leak issues if fault injection happened inside close?

**Review by wangshilong — APPROVED** [review]

File: src/client/dfs/dfs.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfs/dfs.c:1254]: yes, test that covers this and found the array bug is at https://github.com/daos-stack/daos/pull/12502

**Review by liuxuezhao — APPROVED** [review]

File: src/client/dfs/dfs.c (hunk 1 above)
Comment (mchaarawi) [reply src/client/dfs/dfs.c:1254]: correct

**Pull request merged**